working/macros/api/src/protocol/isolate_mirror_impl.dart

```diff
@@ -0,0 +1,249 @@
+import 'dart:async';
+import 'dart:isolate';
+import 'dart:mirrors';
+
+import '../../code.dart';
+import 'protocol.dart';
+import '../../builders.dart';
+import '../../expansion_protocol.dart';
+import '../../introspection.dart';
+import '../../macros.dart';
+
+/// Spawns a new isolate for loading and executing macros.
+void spawn(SendPort sendPort) {
+  var receivePort = ReceivePort();
+  sendPort.send(receivePort.sendPort);
+  receivePort.listen((message) async {
+    if (message is LoadMacroRequest) {
+      var response = await _loadMacro(message);
+      sendPort.send(response);
+    } else if (message is InstantiateMacroRequest) {
+      var response = await _instantiateMacro(message);
+      sendPort.send(response);
+    } else if (message is ExecuteDefinitionsPhaseRequest) {
+      var response = await _executeDefinitionsPhase(message);
+      sendPort.send(response);
+    } else {
+      throw StateError('Unrecognized event type $message');
+    }
+  });
+}
+
+/// Maps macro identifiers to class mirrors.
+final _macroClasses = <_MacroClassIdentifier, ClassMirror>{};
+
+/// Handles [LoadMacroRequest]s.
+Future<GenericResponse<MacroClassIdentifier>> _loadMacro(
+    LoadMacroRequest request) async {
+  try {
+    var identifier = _MacroClassIdentifier(request.library, request.name);
+    if (_macroClasses.containsKey(identifier)) {
+      throw UnsupportedError(
+          'Reloading macros is not supported by this implementation');
+    }
+    var libMirror =
+        await currentMirrorSystem().isolate.loadUri(request.library);
```

Comment:
Can we load a macro from string in memory?

Reply:
I suppose possibly through a `data:` uri? But I haven't tried. This api today actually doesn't even work for `package:` uris although that is just a bug.